crates/hotshot-stake-table/src/vec_based.rs

```diff
@@ -0,0 +1,495 @@
+        let default_bls_comm =
+            VariableLengthRescueCRHF::<F, 1>::evaluate(&to_be_hashed).unwrap()[0];
+        let to_be_hashed = vec![F::default(); STAKE_TABLE_CAPACITY * <K2 as ToFields<F>>::SIZE];
+        let default_schnorr_comm =
+            VariableLengthRescueCRHF::<F, 1>::evaluate(&to_be_hashed).unwrap()[0];
+        let to_be_hashed = vec![F::default(); STAKE_TABLE_CAPACITY];
+        let default_stake_comm =
+            VariableLengthRescueCRHF::<F, 1>::evaluate(&to_be_hashed).unwrap()[0];
+        let default_comm = (default_bls_comm, default_schnorr_comm, default_stake_comm);
+        Self {
+            head: StakeTableSnapshot::default(),
+            epoch_start: StakeTableSnapshot::default(),
+            last_epoch_start: StakeTableSnapshot::default(),
+            head_total_stake: U256::zero(),
+            epoch_start_total_stake: U256::zero(),
+            last_epoch_start_total_stake: U256::zero(),
+            bls_mapping: HashMap::new(),
+            epoch_start_comm: default_comm,
+            last_epoch_start_comm: default_comm,
+        }
+    }
+
+    /// Update the stake table when the epoch number advances, should be manually called.
+    pub fn advance(&mut self) {
+        // Could we avoid this `clone()`?
+        self.last_epoch_start = self.epoch_start.clone();
+        self.last_epoch_start_total_stake = self.epoch_start_total_stake;
+        self.last_epoch_start_comm = self.epoch_start_comm;
+        self.epoch_start = self.head.clone();
+        self.epoch_start_total_stake = self.head_total_stake;
+        self.epoch_start_comm = self.compute_head_comm();
+    }
+
+    /// Set the stake withheld by `key` to be `value`.
+    /// Return the previous stake if succeed.
+    pub fn set_value(&mut self, key: &K1, value: U256) -> Result<U256, StakeTableError> {
+        match self.bls_mapping.get(key) {
+            Some(pos) => {
+                let old_value = self.head.stake_amount[*pos];
+                self.head.stake_amount[*pos] = value;
+                self.head_total_stake -= old_value;
+                self.head_total_stake += value;
+                Ok(old_value)
+            }
+            None => Err(StakeTableError::KeyNotFound),
+        }
+    }
+
+    /// Helper function to recompute the stake table commitment for head version
+    fn compute_head_comm(&mut self) -> (F, F, F) {
+        // Compute rescue hash for bls keys
+        let mut to_be_hashed = self
+            .head
+            .bls_keys
+            .iter()
+            .map(|key| key.to_fields())
+            .collect::<Vec<_>>()
+            .concat();
+        to_be_hashed.resize(
+            STAKE_TABLE_CAPACITY * <K1 as ToFields<F>>::SIZE,
+            F::default(),
+        );
+        let bls_comm = VariableLengthRescueCRHF::<F, 1>::evaluate(to_be_hashed).unwrap()[0];
+
+        // Compute rescue hash for Schnorr keys
+        let mut to_be_hashed = self
```

Comment:
I prefer a more descriptive name like `bls_comm_preimage`, same comments for almost all other variable named `to_be_hashed` in this file

Reply:
changed in acfdc82